### PR TITLE
Refactor cross-tests to use (cty.Value, resource.PropertyMap) pairs

### PR DESCRIPTION
**Pull request by iwahbe**
Stacked PRs:
 * #2502
 * #2503
 * #2501
 * __->__#2500

--- --- ---

### Refactor cross-tests to use (cty.Value, resource.PropertyMap) pairs

Passing `(cty.Value, resource.PropertyMap)` pairs (instead of just passing
`map[string]any` or `tftypes.Value`) is necessary for testing scenarios where the
Terraform value doesn't neatly line up with the Pulumi value, such as #2048.

This commit pushes pkg/tests/cross-tests to work directly with cty.Value and
resource.PropertyMap instead of converting at the last minute.

All tests pass after this refactor. It only effects test code. Existing tests do not need
to be rewritten.

### CODE REVIEW

**Comment by VenelinMartinov** [comment]
This looks like quite a big change in the framework for https://github.com/pulumi/pulumi-terraform-bridge/issues/2048

Do we really want to expose a _third_ way to write cross-tests? That seems like quite a bit of added complexity for a fairly small gain in 2048.

Note that the alternative is to write "manual" cross-tests for edge cases using tfcheck and pulcheck instead of the cross-tests framework. We have existing examples here, I can dig these up. Do we see a big use case for the inputs differing significantly here? Do we have a vision for this going forward?

**Comment by iwahbe** [comment]
> This looks like quite a big change in the framework for #2048

Yeah, I was surprised by the size of the change necessary to simply pass my own values for Pu and TF. As you said, the alternative is a [manual cross-test](https://github.com/pulumi/pulumi-terraform-bridge/blob/263b68d13407f84ccd220329746d8f54df478c02/pkg/tests/schema_pulumi_test.go#L1604-L1635), which takes a surprising amount of code.

> Do we really want to expose a _third_ way to write cross-tests? That seems like quite a bit of added complexity for a fairly small gain in 2048.

[No](https://github.com/pulumi/pulumi-terraform-bridge/pull/2503), I want to decompose cross-tests into building blocks so each use case is supported. The lowest level of this is:

```go
crosstests.Create(tfSchema schema.Resource, tfProperties cty.Value, puProperties resource.PropertyMap)
```

From this, we can easily build back to the functionality that we currently have. #2503 turns `runCreateImputCheck` into a simple wrapper around `Create`, using code we already have written.

P.S. Whats the second way? I know about `runCreateInputCheck`. Is there another method to do this? Did you mean manual re-implementation of cross-tests?

> Note that the alternative is to write "manual" cross-tests for edge cases using tfcheck and pulcheck instead of the cross-tests framework. 

Yes, but we can design cross-tests so we don't need to drop to re-impelemtning our own test framework for special cases.

> Do we see a big use case for the inputs differing significantly here?

Yes. We need to test:

- implicit type conversions
- explicit type changes (#2048 is a start)
- max items one fiddlyness (simulating major version upgrades)
- secret injection (if we ever want to handle secrets, or in Configure)
- etc.

> Do we have a vision for this going forward?

Yes. I can write something up if it would be helpful. 

tldr: I see cross-tests as 2 major components:

1. Given a TF provider, a Pulumi program and a TF program, assert that both programs do the same thing.
2. Infer a Pulumi program from a TF program.

In general, (2) should only be used in service to (1). I want a design where you can use (1) without (2). Coupling (1) and (2) at a technical level not only make it impossible to do some things, but also make the framework harder to understand by conflating concerns.

File: pkg/internal/tests/cross-tests/tf_driver.go [hunk 1]
```diff
@@ -48,17 +48,25 @@ func newTFResDriver(t T, dir, providerName, resName string, res *schema.Resource
...
 }
 
-func (d *TfResDriver) coalesce(t T, x any) *tftypes.Value {
-	if x == nil {
-		return nil
-	}
-	objectType := convert.InferObjectType(sdkv2.NewSchemaMap(d.res.Schema), nil)
-	for k := range objectType.AttributeTypes {
-		objectType.OptionalAttributes[k] = struct{}{}
+func coalesceInputs(t T, schema map[string]*schema.Schema, config any) cty.Value {
```
Comment (t0yv0) [inline pkg/internal/tests/cross-tests/tf_driver.go:51]: This was pre-existent to help construct cty.Value things but is ever-so-slightly sketchy. We need this permanently or temporarily? Why is `config any`? Is this for convenience or it's necessary? If we put some labor in here, could we rewrite all tests to use cty.Value to represent TF-side inputs so these other forms would not need to be supported?

**Comment by VenelinMartinov** [comment]
Discussed offline, my understanding is:
This is still very much an internal tool, we keep it in the internal folder. We are refactoring the code to accept cty.Value, resource.ProeprtyMap as a way to expose what is happening under the hood.

In the general case, test authors are not expected to specify both cty and propertyMaps - there should still be wrappers around this which handle the conversion for you like with the input checks.

This allows us to express both tests which check the same input in TF and pulumi like https://github.com/pulumi/pulumi-terraform-bridge/pull/2503 and tests with slightly different inputs like https://github.com/pulumi/pulumi-terraform-bridge/pull/2502

**Comment by VenelinMartinov** [comment]
We were also wondering why we ended up with tftypes as the common input in cross-tests and then converting tftypes -> cty and tftypes -> propertyMap. Why did we decide against using cty as the common input and doing only cty -> propertyMap?

perhaps @t0yv0 might be able to recover some context here, I don't remember why we ended up with this.

File: pkg/internal/tests/cross-tests/tf_driver.go (hunk 1 above)
Comment (iwahbe) [reply pkg/internal/tests/cross-tests/tf_driver.go:51]: We need this temporarily. If we re-write all tests in `cross-tests` to be written in terms of `cty.Value`, then we can rip this out.

I didn't want to re-write all of our existing tests while iterating on this PR.

Comment (t0yv0) [reply pkg/internal/tests/cross-tests/tf_driver.go:51]: Perfect. Comment on this in-source would be swell.
